processing/src/main/java/io/druid/segment/QueryableIndexStorageAdapter.java

```diff
@@ -760,6 +943,197 @@ public void close() throws IOException
     }
   }
 
+  public static class CursorOffsetHolder
+  {
+    Offset currOffset = null;
+
+    public Offset get()
+    {
+      return currOffset;
+    }
+
+    public void set(Offset currOffset)
+    {
+      this.currOffset = currOffset;
+    }
+  }
+
+  private static boolean isComparableNullOrEmpty(final Comparable value)
+  {
+    if (value instanceof String) {
+      return Strings.isNullOrEmpty((String) value);
+    }
+    return value == null;
+  }
+
+  private static class CursorOffsetHolderValueMatcherFactory implements ValueMatcherFactory
+  {
+    private final ColumnSelectorFactory cursor;
+
+    public CursorOffsetHolderValueMatcherFactory(
+        ColumnSelectorFactory cursor
+    )
+    {
+      this.cursor = cursor;
+    }
+
+    @Override
+    public ValueMatcher makeValueMatcher(String dimension, final Comparable value)
+    {
+      final DimensionSelector selector = cursor.makeDimensionSelector(
+          new DefaultDimensionSpec(dimension, dimension)
+      );
+
+      // if matching against null, rows with size 0 should also match
+      final boolean matchNull = isComparableNullOrEmpty(value);
+
+      final int id = selector.lookupId((String) value);
+      if (id < 0) {
+        return new BooleanValueMatcher(false);
+      } else {
+        return new ValueMatcher()
+        {
+          @Override
+          public boolean matches()
+          {
+            IndexedInts row = selector.getRow();
+            if(row.size() == 0) {
+              return matchNull;
+            }
+            for (int rowVal : row) {
+              if (rowVal == id) {
+                return true;
+              }
+            }
+            return false;
+          }
+        };
+      }
+    }
+
+    @Override
+    public ValueMatcher makeValueMatcher(String dimension, final Predicate predicate)
```

Comment:
It looks like this code won't actually get used right now (since the existing filters would all use row offset matchers). add a comment to that effect?

In the future I assume this would change to support column types other than strings